File: 구름 알고리즘 미션/그래프 탐색(1)/불이야/불이야.py
import sys
from collections import deque
input = sys.stdin.readline

R, C = map(int, input().split())
lab = [list(map(str, input().rstrip())) for _ in range(R)]
directions = [[-1, 0], [1, 0], [0, -1], [0, 1]]

def bfs():
	distance = [[0] * C for _ in range(R)]
	visited = [[False] * C for _ in range(R)]
	q = deque()
	
	for i in range(R):
		for j in range(C):
			if lab[i][j] == '@':
				q.append([i, j])
				visited[i][j] = True
	
	while q:
		cx, cy = q.popleft()
		for dx, dy in directions:
			mx, my = cx + dx, cy + dy
			if 0 <= mx < R and 0 <= my < C and lab[mx][my] != '#' and not visited[mx][my]:
				visited[mx][my] = True
				distance[mx][my] = distance[cx][cy] + 1
				q.append([mx, my])
				
	return distance, visited

dist, visited = bfs()

for i in range(R):
	for j in range(C):
		if lab[i][j] == '&':
			if visited[i][j]:
				print(dist[i][j] - 1)
			else:
				print(-1)

File: 구름 알고리즘 미션/그래프 탐색(1)/불이야/불이야.js
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');

const directions = [[-1, 0], [0, -1], [1, 0], [0, 1]];
const [R, C] = input[0].split(" ").map(Number);
const lab = Array.from({length: R}, () => Array(C).fill(0));
for(let i = 1 ; i <= R ; i++){
	lab[i-1] = input[i].split("");
}

const bfs = () => {
	let visited = Array.from({length: R}, () => Array(C).fill(false));
	let distance = Array.from({length: R}, () => Array(C).fill(0));
	let q = [];
	let head = 0;
	
	for(let i = 0 ; i < R ; i++){
		for(let j = 0 ; j < C ; j++){
			if(lab[i][j] === '@') {
				q.push([i, j]);
				visited[i][j] = true;
			}
		}
	}
	
	while(head < q.length){
		const [cx, cy] = q[head++];
		for(const [dx, dy] of directions){
			const [mx, my] = [cx + dx, cy + dy];
			if(0 <= mx && mx < R && 0 <= my && my < C && lab[mx][my] !== '#' && !visited[mx][my]){
				q.push([mx, my]);
				visited[mx][my] = true;
				distance[mx][my] = distance[cx][cy] + 1;
			}
		}
	}
	
	return [distance, visited];
}

const [dist, visited] = bfs();
for(let i = 0 ; i < R ; i++){
	for(let j = 0 ; j < C ; j++){
		if(lab[i][j] === '&'){
			console.log(visited[i][j] ? dist[i][j] - 1 : -1);
		}
	}
}